### PR TITLE
Williamhe/1900882 add dotnet8 bullseye

### PR DESCRIPTION
**Pull request by william-msft**
- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as:
  - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1900882?src=WorkItemMention&src-action=artifact_link
  - Adding .NET8 support to bullseye.
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

### CODE REVIEW

File: vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh [hunk 1]
```diff
@@ -102,13 +102,13 @@ function copyPlatformBlobsToProdForDebianFlavor() {
...
+    # - platformName=dotnet and debianFlavor=bookworm or debianFlavor=bullseye
     # Not allowed combinations: 
-    # - platformName=python and debianFlavor=bookworm
+    # - platformName=python and debianFlavor=bookworm   
     # - platformName=nodejs and debianFlavor=bookworm
     # - platformName=java and debianFlavor=bookworm
-    # - Any platformName other than dotnet and debianFlavor=bookworm
-    if [ "$platformName" != "dotnet" ] && [ "$debianFlavor" == "bookworm" ]; then
+    # - Any platformName other than dotnet and debianFlavor=bookworm or debianFlavor=bullseye
+    if { [ "$platformName" != "dotnet" ] || ([ "$debianFlavor" != "bookworm" ] && [ "$debianFlavor" != "bullseye" ]); }; then
```
Comment (waliMSFT) [inline vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh:111]: is it not easier to read the code if we tweak the if/else condition like that:

> if { [ "$platformName" == "dotnet" ] && ([ "$debianFlavor" == "bookworm" ] || [ "$debianFlavor" == "bullseye" ]); }; then
       ## copy blobs
else
      ## do not copy blobs

**Review by waliMSFT — COMMENTED** [review]
looks good to me

**Comment by waliMSFT** [comment]
I am assuming that it is the customer's request to add the bullseye dotnet 8.0 image. Please let us know if it is otherwise.

File: vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh [hunk 2]
```diff
@@ -102,13 +102,13 @@ function copyPlatformBlobsToProdForDebianFlavor() {
...
-    # - platformName=dotnet and debianFlavor=bookworm
+    # - platformName=dotnet and debianFlavor=bookworm or debianFlavor=bullseye
     # Not allowed combinations: 
     # - platformName=python and debianFlavor=bookworm
     # - platformName=nodejs and debianFlavor=bookworm
     # - platformName=java and debianFlavor=bookworm
-    # - Any platformName other than dotnet and debianFlavor=bookworm
-    if [ "$platformName" != "dotnet" ] && [ "$debianFlavor" == "bookworm" ]; then
+    # - Any platformName other than dotnet and debianFlavor=bookworm or debianFlavor=bullseye
+    if { [ "$platformName" != "dotnet" ] || ([ "$debianFlavor" != "bookworm" ] && [ "$debianFlavor" != "bullseye" ]); }; then
```
Comment (cormacpayne) [inline vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh:111]: Shouldn't the logic in this case be:

```
if platformName != "dotnet" AND (debianFlavor != "bookworm" OR debianFlavor != "bullseye")
```

Comment (william-msft) [reply vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh]: Yes, thanks for catching!

File: vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh (hunk 1 above)
Comment (william-msft) [reply vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh]: Yes, thanks for catching!

**Comment by william-msft** [comment]
> I am assuming that it is the customer's request to add the bullseye dotnet 8.0 image. Please let us know if it is otherwise.

That is correct.

File: vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh [hunk 3]
```diff
@@ -102,13 +102,13 @@ function copyPlatformBlobsToProdForDebianFlavor() {
...
-    # - platformName=dotnet and debianFlavor=bookworm
+    # - platformName=dotnet and debianFlavor=bookworm or debianFlavor=bullseye
     # Not allowed combinations: 
     # - platformName=python and debianFlavor=bookworm
     # - platformName=nodejs and debianFlavor=bookworm
     # - platformName=java and debianFlavor=bookworm
-    # - Any platformName other than dotnet and debianFlavor=bookworm
-    if [ "$platformName" != "dotnet" ] && [ "$debianFlavor" == "bookworm" ]; then
+    # - Any platformName other than dotnet and debianFlavor=bookworm or debianFlavor=bullseye
+    if [[ "$platformName" != "dotnet" && ( "$debianFlavor" != "bookworm" || "$debianFlavor" != "bullseye" ) ]]; then
```
Comment (cormacpayne) [inline vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh:111]: Wait, sorry, I might have had part of the logic incorrect in my original comment: what exactly are we trying to accomplish with this if-statement? I just want it to be explicitly clear (with examples) what combination of platform versions and Debian flavors should and should not be able to be moved from staging to production?

Comment (cormacpayne) [reply vsts/scripts/publishSdksFromStagingToProdStorageAccount.sh]: OK makes sense to me -- thanks for the explanation

**Review by cormacpayne — DISMISSED** [review]
:shipit:

**Review by cormacpayne — DISMISSED** [review]

File: images/runtime/dotnetcore/8.0/base.bullseye.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+# and copy them to our final runtime image
+FROM mcr.microsoft.com/dotnet/sdk:8.0-preview AS tools-install
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-sos
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-trace
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-dump
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-counters
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-gcdump
+# Update dotnet-monitor after .NET is out of preview to: 
+#   dotnet-monitor --version 8.*
+RUN dotnet tool install --tool-path /dotnetcore-tools dotnet-monitor --version 8.0.0-preview.7.23402
```
Comment (daniv-msft) [inline images/runtime/dotnetcore/8.0/base.bullseye.Dockerfile:11]: Nit: not a blocker for this PR, but we need to remember post GA to update these dotnet-monitor versions. I'll create a work item for it

Comment (william-msft) [reply images/runtime/dotnetcore/8.0/base.bullseye.Dockerfile:11]: Yes we do, thanks Dan!

**Review by daniv-msft — APPROVED** [review]

**Pull request merged**